File: frontend/src/types/monitoring.ts
/**
 * TypeScript definitions for Monitoring, Observability, and Statistical Drift (Phase 13).
 */

export interface HistogramBin {
  bin_label: string;
  baseline_pct: number;
  current_pct: number;
}

export interface FeatureDriftReport {
  feature_name: string;
  feature_type: 'numerical' | 'categorical';
  drift_detected: boolean;
  psi_score: number;
  primary_test: string;
  test_statistic: number;
  p_value: number;
  wasserstein_distance?: number;
  histogram_bins: HistogramBin[];
  baseline_stats: Record<string, any>;
  current_stats: Record<string, any>;
}

export interface ModelDriftAnalysisResponse {
  model_id: string;
  model_name: string;
  model_version: string;
  task_type: string;
  health_status: 'HEALTHY' | 'WARNING' | 'DRIFT_DETECTED';
  drift_score: number;
  drift_percentage: number;
  total_features: number;
  drifted_features_count: number;
  max_psi: number;
  baseline_sample_count: number;
  current_sample_count: number;
  feature_reports: FeatureDriftReport[];
  analyzed_at: string;
}

export interface ModelMonitoringOverview {
  model_id: string;
  model_name: string;
  model_version: string;
  status: string;
  health_status: 'HEALTHY' | 'WARNING' | 'DRIFT_DETECTED';
  total_inferences: number;
  max_psi: number;
  drifted_features_count: number;
  last_inference_at?: string;
}

export interface CustomDriftAnalysisPayload {
  current_dataset_id?: string;
  alpha?: number;
  psi_threshold?: number;
}

File: backend/app/schemas/monitoring.py
"""Pydantic schemas for Model Observability and Statistical Drift Engine."""

import uuid
from datetime import datetime
from typing import Any, Dict, List

from pydantic import BaseModel, ConfigDict, Field


class HistogramBin(BaseModel):
    """Distribution comparison bin."""

    bin_label: str
    baseline_pct: float
    current_pct: float


class FeatureDriftReport(BaseModel):
    """Detailed statistical drift report for a single feature."""

    feature_name: str
    feature_type: str
    drift_detected: bool
    psi_score: float
    primary_test: str
    test_statistic: float
    p_value: float
    wasserstein_distance: float | None = None
    histogram_bins: List[HistogramBin] = Field(default_factory=list)
    baseline_stats: Dict[str, Any] = Field(default_factory=dict)
    current_stats: Dict[str, Any] = Field(default_factory=dict)


class ModelDriftAnalysisResponse(BaseModel):
    """Comprehensive drift analysis response."""

    model_config = ConfigDict(protected_namespaces=())

    model_id: uuid.UUID
    model_name: str
    model_version: str
    task_type: str
    health_status: str  # "HEALTHY" | "WARNING" | "DRIFT_DETECTED"
    drift_score: float
    drift_percentage: float
    total_features: int
    drifted_features_count: int
    max_psi: float
    baseline_sample_count: int
    current_sample_count: int
    feature_reports: List[FeatureDriftReport]
    analyzed_at: datetime


class ModelMonitoringOverview(BaseModel):
    """Model observability overview item."""

    model_config = ConfigDict(protected_namespaces=())

    model_id: uuid.UUID
    model_name: str
    model_version: str
    status: str
    health_status: str
    total_inferences: int
    max_psi: float
    drifted_features_count: int
    last_inference_at: datetime | None = None


class CustomDriftAnalysisRequest(BaseModel):
    """Payload to trigger drift analysis against custom evaluation dataset."""

    current_dataset_id: uuid.UUID | None = Field(
        None, description="Optional evaluation dataset to compare against baseline"
    )
    alpha: float = Field(0.05, ge=0.001, le=0.5, description="Significance level for KS/Chi2 tests")
    psi_threshold: float = Field(0.2, ge=0.01, le=1.0, description="PSI threshold for drift alert")
